source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -440,47 +441,32 @@ Network::FilterStatus Filter::initializeUpstreamConnection() {
   return Network::FilterStatus::StopIteration;
 }
 
-bool Filter::maybeTunnel(const std::string& cluster_name) {
-  if (!config_->tunnelingConfig()) {
-    generic_conn_pool_ =
-        std::make_unique<TcpConnPool>(cluster_name, cluster_manager_, this, *upstream_callbacks_);
-    if (generic_conn_pool_->valid()) {
-      connecting_ = true;
-      connect_attempts_++;
-      generic_conn_pool_->newStream(this);
-      // Because we never return open connections to the pool, this either has a handle waiting on
-      // connection completion, or onPoolFailure has been invoked. Either way, stop iteration.
-      return true;
-    } else {
-      generic_conn_pool_.reset();
-    }
+bool Filter::maybeTunnel(Upstream::ThreadLocalCluster& cluster, const std::string& cluster_name) {
+  GenericConnPoolFactory* factory = nullptr;
+  if (cluster.info()->upstreamConfig().has_value()) {
+    factory = &Envoy::Config::Utility::getAndCheckFactory<GenericConnPoolFactory>(
+        cluster.info()->upstreamConfig().value());
   } else {
-    auto* cluster = cluster_manager_.get(cluster_name);
-    if (!cluster) {
-      return false;
-    }
-    // TODO(snowp): Ideally we should prevent this from being configured, but that's tricky to get
-    // right since whether a cluster is invalid depends on both the tcp_proxy config + cluster
-    // config.
-    if ((cluster->info()->features() & Upstream::ClusterInfo::Features::HTTP2) == 0) {
-      ENVOY_LOG(error, "Attempted to tunnel over HTTP/1.1, this is not supported. Set "
-                       "http2_protocol_options on the cluster.");
-      return false;
-    }
-
-    generic_conn_pool_ = std::make_unique<HttpConnPool>(cluster_name, cluster_manager_, this,
-                                                        config_->tunnelingConfig()->hostname(),
-                                                        *upstream_callbacks_);
-    if (generic_conn_pool_->valid()) {
-      generic_conn_pool_->newStream(this);
-      return true;
-    } else {
-      generic_conn_pool_.reset();
-    }
+    factory = &Envoy::Config::Utility::getAndCheckFactoryByName<GenericConnPoolFactory>(
+        "envoy.filters.connection_pools.tcp.generic");
```

Comment:
getAndCheckFactoryByName may throw exception on invalid type_config 

Is such "non-exist" factory config validated by XDS already?

Reply:
Nice catch.  This does look like data plane code with no obvious try/catch blocks in sight.  The most concerning one is the getAndCheckFactory on line 447 above which is a function of config.

Reply:
I think it's safe to lookup by name _providing_ you can guarantee that the filter will be there; it will throw otherwise. The config thing becomes an issue only when you are actually loading config.

I wonder if we can make this more robust going forward by having the factory functions ASSERT that they are in a control plane context, using some main thread singleton. There's probably static annotations etc. we could use if we wanted to get fancy, but I can see how to build this dynamic check.

Also, Quic has this same pattern.

Reply:
Calling the factory lookup functions in a non control plane setting is not legal.  I urge we invest in adding the appropriate ASSERTs.

Reply:
refactored to make it non-fatal and added a test.  PTAL!